ring/model.go

```diff
@@ -115,6 +121,24 @@ func (d *Desc) IsReady(now time.Time, heartbeatTimeout time.Duration) error {
 	return nil
 }
 
+func (d *Desc) IsReadyZoneAware(now time.Time, heartbeatTimeout time.Duration, zone string) error {
+	numTokens := 0
+	for _, instance := range d.Ingesters {
+		if err := instance.IsReady(now, heartbeatTimeout); err != nil {
```

Comment:
I think it makes more sense to skip iteration on all errors from the same zone. We only need to check states on instances from different zones (this will also skip heartbeat errors in the same zone, which is fine IMO).